analysis/README.md

```diff
@@ -0,0 +1,623 @@
+# Analysis
+
+This crate is where most of the compilation pipeline lives. It takes a parse tree and returns a tree of constrained machines.
+
+## Definitions
+
+We define two types of machines: virtual machines and constrained machines. Constrained machines are the lower level kind of machine. They have a notion of blocks through a latch and a function_id. Virtual machines are a higher level type of machines. For each machine type, we provide the elements which only appear exclusively in that type.
+
+|                         | Virtual | Constrained |
+|-------------------------|---------|-------------|
+| pc                      | yes     | no          |
+| latch                   | no      | yes         |
+| function_id             | no      | yes         |
+| functions bodies        | yes     | no          |
+| internal instructions   | yes     | no          |
+
+The pipeline accepts both kinds of machines, and they are represented by the same type `Machine`. Some steps are specific to virtual machines. They can still be applied to constrained machines and must have no effect. In the process, virtual machines get reduced to constrained machines by encoding their high-level elements into constrained machines elements.
+
+## Pipeline
+
+We go through the pipeline using the following program as an example:
+
+```
+machine Main {
+
+    degree 16;
+
+    DifferentSignatures sub;
+
+    reg pc[@pc];
+    reg X[<=];
+    reg Y[<=];
+
+    instr identity X -> Y = sub.identity
+    instr one -> Y = sub.one
+    instr nothing = sub.nothing
+
+    function main {
+        nothing;
+        return;
+    }
+}
+
+// A machine exposing functions of different signatures
+machine DifferentSignatures {
+
+    reg pc[@pc];
+
+    function identity x: field -> field {
+        return x;
+    }
+
+    function one -> field {
+        return 1;
+    }
+
+    function nothing {
+        return;
+    }
+}
+```
+
+### Type checking
+
+Type checking takes a parse tree and returns a tree of machines. The output type aims at being as strict as possible.
+
+### Virtual machine reduction
+
+Virtual machine reduction turns virtual machines into constrained machines. It has no effect on contrained machines.
+
+#### Inference
+
+Inference infers the value of assignment registers.
+
+```diff
+-               A <== one();
++               A <=Y= one();
+```
+
+#### Batcher
+
+The batcher groups statements within each function into batches which can be executed in the same execution step. In our example, we gain 33% in the main machine by batching the first label with the following statement.
```

Comment:
which following statement?

Reply:
Fixed, didnt have the latest version of the example in here

Reply:
I still don't get it. The next diff is still the END BATCH comments only.

Reply:
Yes, the batcher only inserts the batch delimitations. It does not change the statements.

Reply:
Ah, I think I misunderstood the text. So "the following statement" refers to the statement after said label, right? I imagined a statement after the sentence